accepted/future-releases/language-versioning/package-config-file-v2.md

```diff
@@ -0,0 +1,134 @@
+
+For the *language versioning* feature, we have proposed extending the format by adding fragments to the URI references and an entry with the empty string as key. This is a breaking change (the empty key was not previously accepted) and even more breaking for tools which assume the values are paths.
+
+Instead of breaking the existing file users, we will instead introduce a *new* file containing the information that we intended to store in the `.packages` file. This fill will use a (more) structured format which allows extending it with further information in the future. We will still generate the `.packages` file for now, allowing users of the file to continue working until they can migrate to using the new file.
+
+## New File Contents
+
+The file will contain the information that was planned for the `.packages` file:
+
+- A *default package* specified by its package name.
+- A list of *packages*, each with:
+  - A *package name*.
+  - A *path* or *location URI*.
+  - A *language version* (major/minor version number)
+
+On top of that, we also add extra metadata which allows us to version the format and attach information about when and how the file was created.
+
+We use JSON for the format. JSON has parsers available for most languages (not all Dart tools are written in Dart) and it is standardized and efficiently parsable.  JSON is *not* easy to write by hand (but doable), and it does not allow comments, but this is a machine generated file that will hardly ever be edited by hand.
+
+## New File Format
+
+The new file will be called `.packages.json` and stored in the same directory as the `.packages` file. The format is JSON text describing is a single JSON object.
+
+That object has, at least, the following properties:
+
+- `apiVersion`: A single integer marking the version of this format, currently `2` (counting `.packages` as the first). Tools should refuse to work on files with a larger version number than the ones they are aware of. Incremental changes will not require increasing the version, so new versions are not expected any time soon, but code should still prepare for it.
+
+- `defaultPackage`: A string containing a package name.
+- `packages`: A list of packages, each with the following properties:
+  - `name`: A string containing a package name.
+  - either `path`: A string containing a relative or absolute path using `/` as separator. Every character other than `/` is used verbatim. If the path starts with a `/` or with a DOS/Windows drive letter (a single letter followed by `:/`), then it is considered absolute. If not, the path is relative to the directory of the `.packages.json` file. 
```

Comment:
I really think we don't want to be in the business of defining relative and absolute paths in this spec. (For example, what about UNC paths on Windows? Is a path starting with a drive letter considered an absolute path on non-Windows systems? Should it be?)

Do we need to support paths at all or can we get away with just URIs? They've already done the hard work of specifying relative and file URIs.

Reply:
One of the problems we had with the original format is that Pub always generated URI references that were just relative (to the file) or absolute (pointing to the pub cache) paths. 3rd party tools would read and understand them as plain *paths*, not URI references. They were apparently completely successful, so it seems it is all we actually need.

We an say that the "path" is a *URI path*, but we risk users *still* treating it as a plain path because in 99.99% of the the cases it's the same thing.
The last case is, say, when the pub cache in `C:\Users\Lasse Nielsen\LocalData\.pubcache`, and the path entry for a package becomes `C:/Users/Lasse%20Nielsen/LocalData/.pubcache/mypkg-1.0.0/lib`. Any tool using that as a plain OS path will break because `Lasse%20Nielsen` is not a directory.

So, we have to say whether the content is a URI reference, a URI path, or an OS Path. There are cases where the three differ, and using one a another will give the wrong result, fx  `c:/something`, relative to `file:///c:/src/mypkg/.dart_data/something.json`.
- If treated as URI, `c` is a scheme, and the result is the URI `c:/something`
- If treated as URI path, it is a *relative* path, and the result is file:///c:/src/mypkg/.dart_data/c:/something`
- if treated as OS path, it's an absolute path (and likley the intended approach).

We need to make sure that we specify *exactly* how to interpret the content, and knowing that 3rd party readers may treat it as a OS path, we could just say that it is one.
If we do that, then we should probably not begin defining what's relative and what's not, and leave that to the OS itself (or code that knows which OS it's on).

I don't like the "URI path" approach. It's cheating, really, and just increases the risk of someone treating it as a path, not a URI path and getting caught by `%20` escapes or similar.

So, the other reasonable approach is to say that it is a URI reference. In that case, it should be named `uri` (just for explicitness), and I'd like if Pub wrote `file://` in front of absolute paths (maybe all of them, maybe some of them, just to keep people on their toes), just to make sure that people who think everything is a path gets caught.

So, please vote :+1: for OS path, :heart: for URI reference :smile: .

Reply:
One key problem with OS paths is that they are not context free. You don't know what the string "foo\bar" means unless you know which host operating system it is intended for. That alone to me is a big enough problem that I think it's better to say these are URIs.

> So, the other reasonable approach is to say that it is a URI reference. In that case, it should be named uri (just for explicitness), and I'd like if Pub wrote file:// in front of absolute paths (maybe all of them, maybe some of them, just to keep people on their toes), just to make sure that people who think everything is a path gets caught.

+1 to all that, including naming it "uri".

Reply:
I tend to agree. Saying that it is a URI reference will make things easier. Even if I say that the path uses `/` as separator, it won't prevent `"ab/cd\ef"` from being an interpretation nightmare.

So, yes, URIs. Not paths, because paths are not standardized.